Add Node "editor-only" Toggle

torcado194 opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/7715
Related: https://github.com/godotengine/godot-proposals/issues/3433

This PR adds editor functionality to toggle "editor-only" mode for Nodes in the scene tree, allowing you to omit (or "comment out") nodes when running or exporting the project. This is especially useful for prototyping, as it allows users to quickly test and compare different nodes/scenes.

This is implemented via editor metadata, which the scene packer checks for to replace relevant nodes with empty Node objects when building. This prevents node path references from erroring.

The editor functionality is contained in a context menu option, which is assignable to a hotkey for quick access.

![demo2](https://i.imgur.com/qtXrw28.gif)

<details><summary>Original description</summary>
<p>

This PR adds an additional toggle to nodes in the tree, allowing you to toggle the existence of the node when running or exporting the project. This allows for **editor-only** nodes and **completely removed** nodes, when combined with the existing visibility toggle.

This is one of the first changes I made to Godot locally, about a year ago. I find it invaluable for prototyping. I've since given the code to friends and anyone who's asked for it. I figured enough people have wanted it that I should submit a PR for it.

![preview](https://i.imgur.com/VVMbIMv.gif)

(This could alternatively be hidden behind the right-click menu. I opted for this solution since I primarily use it for quick prototyping, so having immediate access is preferred.)

Note about the implementation:
I originally implemented this as an additional `tag` on the node resource, so that when parsing the node in `resource_format_text.cpp` it could apply the `disabled` state directly so that `SceneState::instantiate()` could read that data up front rather than needing to parse the properties. The metadata implementation was used instead because I did not want to interfere with the binary saving/loading system, which is necessary for keeping the behavior when exporting. I still believe the tag is the better solution, if anyone has insights about this I would appreciate it.


</p>
</details>

torcado194 committed:
node existence toggle

AdriaandeJongh commented:
In terms of UX, this is another icon in a place that is already crowded with icons. A node can have a whopping 5 icons when it's in a group, has a connected signal, has a script, is visible, and now is enabled.

passivestar commented:
Hey, I like it

I have some questions:

1. Would it make sense for this button to also toggle visibility in editor? If editor and runtime mirror each other you'll always know what's gonna happen when you press play and there will be less _"where did my stuff go? oh that's right I disabled it"_ moments. It's not immediately obvious what the value of seeing disabled stuff in the viewport is. I know you can make it a habit to click both, but it's strictly a UX question
2. Does this also disable editor `@tool`s? If I have a disabled tool node saved as a scene, will its `_enter_tree()` run when I instantiate it in editor?
3. Are you sure about the diamond shape? It's often associated with keyframing in software. What do you think of a rounded square instead?

yoont4 commented:
Would really like to see this go in or something similar. I've been wanting an easy way to test scenes with nodes on/off without having to completely destroy it and lose all data, or make any editor plugin hacks to work around it.

BliznyukNM commented:
I was thinking about UX problem @AdriaandeJongh mentioned. One solution I can think of is having 3-state visibility button - visible, hidden, disabled (as disabled is hidden at the same time). Eye icon in hidden state can be filled half-way to illustrate that there is yet another click possible.

And to go back to visible state again you would have to cycle through all states.

AdriaandeJongh commented:
Does the icon really need to be in the SceneTree for quick access? I can imagine it is initially enough to having it in the inspector together with the Node properties. If then enough people use the feature and want 'quick access' to it, then a followup PR could add the icon.

grgp commented:
I think a check icon just like in Blender/Unity would work best? In Blender, it's also possible to filter the icons that you want to see on the right, but that's probably a separate PR.

<img width="279" alt="image" src="https://github.com/godotengine/godot/assets/12206156/3435519a-0697-4863-8833-49d2e08efd5e">

<img width="279" alt="image" src="https://github.com/godotengine/godot/assets/12206156/3734227a-2d62-4164-95aa-d989b876c735">

For now I think I agree with AdriaandeJongh that for now it can be placed at the very top of the inspector, just like in Unity. I think this feature is direly needed, so the faster it can be merged in, the better.
<img width="285" alt="image" src="https://github.com/godotengine/godot/assets/12206156/4ef82d6d-b466-4eda-ae0e-602f2691e5d8">

KoBeWi commented:
See also #56446

AThousandShips commented:
How does this close https://github.com/godotengine/godot-proposals/issues/3433? I don't see any editor only or game only options for this.

LauraWebdev commented:
Having this as a quick button in the SceneTree would be a game changer for my productivity. I personally enjoy the "three state button" idea shared to combine visibility and activity, although there are also valid reasons to not combine them (one being a changed workflow for those not interested in this feature)

AThousandShips commented:
@LauraWebdev Please don't bump without contributing significant new information. Use the :+1: reaction button on the first post instead.

torcado194 commented:
@AdriaandeJongh yep, like mentioned in the post this could alternatively be hidden under the right-click menu to reduce the clutter. personally, i prefer it immediately accessible (for prototyping, etc.). I don't have any strong opinions one way or the other, this was simply what I ended up preferring.
IMO, I think this should be the default behavior of the eye icon, and disabling drawing should be hidden away. I consider making a node invisible but still active a more niche use case than toggling its existence entirely. But that's a much more radical change that I wouldn't expect to be considered :P 

@passivestar 
1. Part of the point of this feature is that it enabled **editor-only** nodes. I can use this to place visual markers and indicators in scenes and have them not exist in the built project. If you want something completely removed, you disable both icons, they work in tandem. Understandably there's a disconnect here between what you see in-editor and in-game, but that's the point. otherwise, like i said, this shouldn't be a separate toggle and should replace the eye icon behavior. Changing this would remove 50% of the usability for me.
2. Yes, it doesn't impact node's state in the editor, only when building. If you disable existence on a `@tool` node, it simply removes it from the built project. It still exists and runs in-editor.
3. I don't have any strong connection to the icon. It's simply what I ended up using. There's tons of discussion about the UI design in https://github.com/godotengine/godot-proposals/issues/7715, but no consensus was made. Personally, I think merging this behavior into the eye icon (so that it has 3+ states) is clever but unnecessarily confusing for people not understanding what it does. However, if this feature were hidden behind the right-click menu, I think it would be nice to have this existence toggle controlled by ctrl+clicking the eye icon (or another modifier key). 
Here's a demo:
![demo](https://i.imgur.com/OFygYVL.gif)

@AThousandShips By disabling existence and not visibility, a node remains visible and active in the editor but not in the built project. This is done automatically. However, this doesn't implement the inverse behavior (the node exists in the built project but not in the editor). So I don't believe this fully closes that proposal. I've edited the main post.

passivestar commented:
@torcado194 What about renaming "Toggle Existence" to "Toggle Editor-Only" to make it unambiguous?

torcado194 commented:
@passivestar I think that's fine, yeah. It didn't register to me that that is what this is ultimately doing.

Alternatively, I've considered the idea of a 3-state existence toggle: "editor-only", "game-only", and "disabled". This would end up closing https://github.com/godotengine/godot-proposals/issues/3433, maybe as a separate PR. But I do think I want to implement this for myself now haha

passivestar commented:
> if this feature were hidden behind the right-click menu, I think it would be nice to have this existence toggle controlled by ctrl+clicking the eye icon

What if:

- `Click` - Toggle visibility
- `Shift + Click` - Toggle editor-only
- `Ctrl + Click` - Toggle both (if they are in different states ctrl+clicking would first bring them to the same state based on the new visibility state)

This would make both workflows one-click

AdriaandeJongh commented:
Toggling visibility and toggling whether a node is editor-only are two wholly different concepts to me. On top of that and unlike toggling visibility, making a node editor only has a lot of far reaching consequences such as non-editor-only nodes not being able to reference them. Such consequences make me vote strongly against putting it under / alongside the visibility toggle.

I like the feature / idea a lot, but my suggestion would be to put the option to make a node editor-only under the Node category, alongside the fields for process and metadata and such.

torcado194 commented:
@AdriaandeJongh I think those points are fair. I feel like these different solutions are circularly conflicting (having a button on the node tree allows for quick prototyping but adds clutter, removing the button but adding a shortcut to toggle it removes the clutter but semantically couples "editor-only" and "visibility" when they should be distinct, moving the button to the inspector decouples the properties but removes the quick toggle).

I prefer some way to quickly toggle the state because prototyping is my main use case for this feature, but if the button were moved to the inspector panel it wouldn't really matter to me, i'd just make a plugin to add a hotkey/button to switch it anyway. Whatever the consensus is, I'm happy to make changes :)

passivestar commented:
> I prefer some way to quickly toggle the state because prototyping is my main use case for this feature

I think having this function in the context menu and exposed to editor's shortcuts to be able to bind some hotkey to it (can be empty by default) is the best compromise here. Then you don't need to make an addon

AThousandShips commented:
Let's discuss the more general details of the feature on the proposal, and strictly the implementation details here

Snowdrama commented:
> Personally, I think merging this behavior into the eye icon (so that it has 3+ states) is clever but unnecessarily confusing for people not understanding what it does. 

> I've considered the idea of a 3-state toggle: "editor-only", "game-only", and "disabled". 

@torcado194 For me the red text indicates that there's an "issue" with the node so I personally would avoid that just because for me it indicates that I need to check it to see if something is broken, like a missing resource reference or something.

I do think it needs some kind of visual that clearly shows it's somehow different than other nodes without looking like the node is somehow broken or has an error. 

If the icon has multiple states, I think it would make sense to change the icon itself as well as the hover tooltip. So for example changing the eye to a dark grey E that says "Editor-Only" on hover or something similar, while the eye icons continue to say "Toggle Visibility" on hover. 

> What if:
> 
> * `Click` - Toggle visibility
> * `Shift + Click` - Toggle editor-only
> * `Ctrl + Click` - Toggle both (if they are in different states ctrl+clicking would first bring them to the same state based on the new visibility state)

@passivestar Adding a shortcut to speed up toggling it would definitely be nice 100%, my only thing would be that the functionality would be hidden behind a shortcut and the information for it isn't surfaced anywhere in the editor itself which I think is a valuable even if there is a shortcut. I think that would be solved by @AdriaandeJongh's suggestion of putting it in the Inspector with the Process drop down, a tooltip could be put there saying something like "This can also be toggled by Ctrl+Clicking the Visibility toggle icon in the scene view" or something. 

> I like the feature / idea a lot, but my suggestion would be to put the option to make a node editor-only under the Node category, alongside the fields for process and metadata and such.

@AdriaandeJongh I personally like having it in the scene tree because being able to tell what is set to editor-only is useful to be able to at a glance see what won't be added to the build. Putting the toggle in the inspector view with process would be fine, but I would still have some kind of visual indicator to show that it is editor-only in the scene view.

Kalto-Mate commented:
This is absolutely needed, but the whole icon debate is making another problem surface: There's already a LOT of icon clutter and it's a creeping problem, so one thing that does come to mind is to have the visibility toggle house node disabling functionality. 

Basically, you would cicle from visible to hidden to disabled, then back to visible. Maybe eye, closed eye and forbidden symbol. My reasoning is that I don't see how anyone would need to make a node invisible if they are going to also disable it. Double clicking MIGHT get annoying, so maybe other mouse buttons could cicle backwards if used on top of the button.

I know I'm glossing over the disabling scope nuance (editor only, game only) but I feel it might be justifiable for this to be buried just under the surface instead of a button on the tree. After all is an specific and deliberate technical choice and you really don't want to set stuff to editor only on accident and then be confused. To that extent a node set to be editor only could rock a "G" within a forbidden sign instead of the plain forbidden sign with nothing inside (which means disabled across the board)

![image](https://github.com/godotengine/godot/assets/74378599/ae5df9b5-62e7-4e98-b883-fd61b2fdee36)

PLyczkowski commented:
I would go with a dropdown menu at the top of the inspector with Enabled/Disabled/Editor-only (RMB context menu is also fine but less discoverable), not adding a clickable icon, and signifying state in the scene tree view with graying out the name for disabled and coloring it for editor-only (I think yellow is taken, so maybe blue?)

torcado194 commented:
I have made changes following the suggestions made here and in the linked proposal.

- Removed the icon from the scene tree
- Added a button to the context menu of tree items to toggle editor-only (which also allows for an optional shortcut for the toggle)
- Marked editor-only nodes as *blue* (and descendants faded)
- Changed implementation to use `editor_only` in place of `disabled` internally
- Fixed formatting

I think it also makes sense to rename this PR to "Add Editor-Only Toggle", let me know if there's any reason I shouldn't do that.

Here is a demo of the changes:
(The shortcut is empty by default, I added one just for the demo)

![demo2](https://i.imgur.com/qtXrw28.gif)

Kalto-Mate commented:
I fear using color to convey information, SPECIALLY blue is a big no-no for accesible design. I have no idea how much control over the text formatting on node's names is exposed, but would bold, italics, crossed-through or even color flipped presentation (see image) be more prudent?

![623tge7289](https://github.com/godotengine/godot/assets/74378599/d6e3b4c4-a822-40be-a522-d21b9f0d49ea)

On a side note, maybe it's just me, but back when I used Unity I did some stuff by having their equivalent of nodes disabled full stop and re-enabling them at runtime, so settling with just editor only disabling does narrow possibilities a bit.

Calinou commented:
> I have no idea how much control over the text formatting on node's names is exposed, but wouldn't bold, italics, crossed-through or even color flipped presentation (see image) be more prudent?

To my knowledge, you can't use different font styles for individual tree items, but you can give them a dedicated background color. This is already done in the FileSystem dock for colored folders.

KoBeWi commented:
> To my knowledge, you can't use different font styles for individual tree items

*Technically* you can use custom draw for that and draw the text manually.

Kalto-Mate commented:
> I have made changes following the suggestions made here and in the linked proposal.

Just curious, is hijacking the eye button like I suggested really far out there and off the table?

Macklehatton commented:
> Just curious, is hijacking the eye button like I suggested really far out there and off the table?

Look at #7715 and #3433 for context and more theoretical discussion.

PLyczkowski commented:
Oh, blue is already used for editor-running script icons, did not notice that. I guess it either makes blue a better candidate for editor-only nodes from this patch, or if it is rejected due to accessibility then the editor-running script icons should be looked at as well (they could have a T letter on the icon for example).

![image](https://github.com/godotengine/godot/assets/7307968/50c8e0d6-2ff4-4d0f-a12f-52cb2ded9fb7)

passivestar commented:
@PLyczkowski it's not blue it's the accent color:

<img width="400" alt="image" src="https://github.com/godotengine/godot/assets/60579014/46ff29ca-0b7c-4aa3-8e4d-f58fe3673d7b">

PLyczkowski commented:
> @PLyczkowski it's not blue it's the accent color

Oh. Not sure that's a good implementation then, no way to remember 'tool scripts look like this'... Icon with added gear or the letter T would be better imo

adamscott commented:
Sorry for the delay.

I love the idea of a node existence toggle, but I don't really like the fact that it's closely binded to the editor.
In my mind, it should be possible to do this from code too. It would add complexity to the feature, though.

Kalto-Mate commented:
> Sorry for the delay.
> 
> I love the idea of a node existence toggle, but I don't really like the fact that it's closely binded to the editor. In my mind, it should be possible to do this from code too. It would add complexity to the feature, though.

As I understand it, this proposal gets the job done by doing some hacky procedures to get nodes to actually cease to exist. The moment disabling becomes an official clean feature, like a boolean you can set, both code use and gui implementation will be easier.

Snowdrama commented:
> As I understand it, this proposal gets the job done by doing some hacky procedures to get nodes to actually cease to exist. 

Based on what I've read it adds a new value to the base node class that when set, makes it so the routine to instantiate a packed scene doesn't actually instantiate the nodes marked as "editor-only" if the scene is being loaded in a build.

>  it should be possible to do this from code too

I think adding code features could be added later, simply because adding the "editor-only" flag at runtime would not have any effect since this affects the Packed Scene loading routine. So making it accessible via code would only add the ability for tool scripts to set the flag in the editor, and while that could be useful, I don't think it should be a blocker.

There's another proposal that is discussing something that is different, since the node is loaded but can be "enabled/disabled" I think there it makes way more sense since the goal is to make it able to be toggled at runtime:
https://github.com/godotengine/godot-proposals/issues/7715

The only other additional code driven enhancement might also be an additional flag in GDScript like `@editor_only` that is similar to how `@tool` works, and would just force on the "editor_only" flag on assignment, but this wouldn't get rid of the ability to set it in the editor, and I don't think this should inherently be a blocker either.

conde2 commented:
> I think a check icon just like in Blender would work best? In Blender, it's also possible to filter the icons that you want to see on the right, but that's probably a separate PR.
> 
> <img alt="image" width="279" src="https://private-user-images.githubusercontent.com/12206156/333865474-3734227a-2d62-4164-95aa-d989b876c735.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyOTE4MDAsIm5iZiI6MTcxOTI5MTUwMCwicGF0aCI6Ii8xMjIwNjE1Ni8zMzM4NjU0NzQtMzczNDIyN2EtMmQ2Mi00MTY0LTk1YWEtZDk4OWI4NzZjNzM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDA0NTgyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg0ZWY4MDkyMDRkYzdlZjIwMDEwYzUzYjFiYmZiOTY0ZDk4M2IwMTAzNjA2YzZkYzA3NDAxMzAzODVhNmJkNWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zOAjFFSDu37CML6xdhqy3r5Dl0S3FNekVtBostzIybs"> <img alt="image" width="279" src="https://private-user-images.githubusercontent.com/12206156/333865461-3435519a-0697-4863-8833-49d2e08efd5e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyOTE4MDAsIm5iZiI6MTcxOTI5MTUwMCwicGF0aCI6Ii8xMjIwNjE1Ni8zMzM4NjU0NjEtMzQzNTUxOWEtMDY5Ny00ODYzLTg4MzMtNDlkMmUwOGVmZDVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDA0NTgyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5OWRkNjMyNzc2M2Q1YWRiZjdjMTJiY2U0M2JjNzkwN2YxNTRiNzVmNjUwNTA0MWRlYTM2Mzk2ODkwN2U0ZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KcHQ4NkvP9Wz7zoHvU18TlHI_tY48KskOHzSh9xbZm4">
> For now I agree with AdriaandeJongh that for now it can be placed at the very top of the inspector, just like in Unity. <img alt="image" width="285" src="https://private-user-images.githubusercontent.com/12206156/333865599-4ef82d6d-b466-4eda-ae0e-602f2691e5d8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyOTE4MDAsIm5iZiI6MTcxOTI5MTUwMCwicGF0aCI6Ii8xMjIwNjE1Ni8zMzM4NjU1OTktNGVmODJkNmQtYjQ2Ni00ZWRhLWFlMGUtNjAyZjI2OTFlNWQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDA0NTgyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQyOGJhOGVlZmQ4OWFhMmZmNTAxMTI5Nzk4Y2RiNmY5MTgxNGU2N2FhZmEzZGQzN2Y3YWMyN2EyNTMzOGE5Y2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ABZvuRZWg1pVN4dcoBEhT__qadnmNuF3lWxiWZyZUF4">
> 
> Also worth noting that if you disable a gameobject in Unity, the text on the scene tree will be greyed out. <img alt="image" width="226" src="https://private-user-images.githubusercontent.com/12206156/333865896-646b5ca1-93a3-4dce-b464-8cf65fde7cf0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyOTE4MDAsIm5iZiI6MTcxOTI5MTUwMCwicGF0aCI6Ii8xMjIwNjE1Ni8zMzM4NjU4OTYtNjQ2YjVjYTEtOTNhMy00ZGNlLWI0NjQtOGNmNjVmZGU3Y2YwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDA0NTgyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwYmI3MGFjY2I5ZDIyMzhkZGE0Y2JhYzkxODQ4OTMxMzM2ZmFkN2UxMDJiMTIyMWUzMzM5ZmUyZDAwOGJmMmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Q1DMt8gepMV65kiuDqgyn3i6ZEqmgMoCdj1igHQoc-4">
> 
> I think this feature is direly needed, so the faster it can be merged in, the better.

I also prefer this solution over the others persented here, having it in the inspector with a simple checkbox is enought, clean and easy to access.

geekley commented:
I think whatever implementation is used for an "editor-only" feature, it should be at least forward-thinking about possible future extension to a more general "availability" feature, of which "editor-only" is just a subset. Plus "editor-only" is not specific enough, as you may want different definitions for it.

So ideally we should be able to set any availability, to define when nodes are included or excluded completely from builds. Luckily there's already a feature tags concept, which should be perfect for this, e.g.:
- this node only exists in the editor when not running: `metadata/availability = "editor_hint"`
- this node only exists at runtime (including when running from the editor): `metadata/availability = "!editor_hint"`
- this node only exists at runtime when running from the editor: `metadata/availability = "editor_runtime"`
- this node only exists on the editor (whether or not it's running): `metadata/availability = "editor"`
- this node only exists on exported games: `metadata/availability = "template"`
- this node only exists on mobile platforms: `metadata/availability = "mobile"`
- this node only exists on (exported?) desktop/laptop platforms: `metadata/availability = "pc"`
- this node only exists on web platform: `metadata/availability = "web"`
- this node only exists on my custom feature tag (e.g. per-store builds): `metadata/availability = "google_play"`

Basically, it should set a feature tag representing the builds where the node can be included, or a preceding `!` to include it everywhere except on builds with that tag. I'm talking about how this is serialized -- editor UI could be just a simple toggle for `"editor_hint"` VS `""` (everywhere) for now, and implementation expanded later.

This is a flexible implementation, since you could also in the future allow maybe even expressions with `|` and `&` in the string, e.g. `"pc & !editor"`, and its format would be compatible.

Also, this concept could be later expanded to resources as well.

AThousandShips commented on scene/resources/packed_scene.cpp:
```diff
@@ -336,6 +350,15 @@ Node *SceneState::instantiate(GenEditState p_edit_state) const {
+					if (!Engine::get_singleton()->is_editor_hint()) {
+						if (snames[nprops[j].name] == "metadata/_editor_only_" && props[nprops[j].value]) {
+							editor_only = true;
+							node = Object::cast_to<Node>(memnew(Node));
```
```suggestion
							node = memnew(Node);
```

AThousandShips commented on scene/resources/packed_scene.cpp:
```diff
@@ -191,7 +191,19 @@ Node *SceneState::instantiate(GenEditState p_edit_state) const {
+			}
+		}
+		if (editor_only) {
+			node = Object::cast_to<Node>(memnew(Node));
```
```suggestion
			node = memnew(Node);
```

LauraWebdev commented:
> Unity's implementation is a bit different btw - Unity’s GameObject.SetActive(bool) also toggles a GameObject’s active/inactive state at runtime or play mode, stopping execution and hiding it - they stop receiving Update(), physics, rendering, and event callbacks.
> 
> This has an additional benefit over the implementation proposed here that one can toggle active state of a game object and it's children at runtime for testing purposes. Not sure if it's viable to implement it in Godot like that due to it's node based design, as opposed to the Unity's game objects + components, but one can simulate the behavior by calling:
> 
> set_process(false) set_physics_process(false) visible = false set_block_signals(true)
> 
> But there’s no single “active/inactive” flag that halts all of them simultaneously.
> 
> In summary: This feature would become even more useful if it allowed for runtime toggling.

In a way, the "Godot" way would probably look like a "process_mode" of "editor_only"? I'm following this issue for a while now and just thought of this.

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -3933,50 +3828,42 @@ void SceneTreeDock::_tree_rmb(const Vector2 &p_menu_pos) {
+		}
 	}
 
+	menu->add_icon_shortcut(get_theme_icon(SNAME("GuiExistenceDisabled"), SNAME("EditorIcons")), ED_GET_SHORTCUT("scene_tree/editor_only"), TOOL_EDITOR_ONLY);
```
```suggestion
	menu->add_icon_shortcut(get_editor_theme_icon(SNAME("GuiExistenceDisabled")), ED_GET_SHORTCUT("scene_tree/editor_only"), TOOL_EDITOR_ONLY);
```

AThousandShips commented on scene/resources/packed_scene.cpp:
```suggestion
			node = memnew(Node);
```
Redundant

AThousandShips commented on scene/resources/packed_scene.cpp:
```suggestion
							node = memnew(Node);
```
